src/ode23s.jl

```diff
@@ -0,0 +1,153 @@
+    dt   ::T
+    F1   ::Vector{Y}
+    F2   ::Vector{Y}
+    J    ::Matrix{Y}
+    iters::Int
+end
+
+
+# for debugging
+function show(io::IO, state::RosenbrockState)
+    show(io,state.step)
+    println("dt =$(state.dt)")
+    println("F1 =$(state.F1)")
+    println("F2 =$(state.F2)")
+    println("J  =$(state.J)")
+end
+
+
+function start{O<:ExplicitODE,S<:ModifiedRosenbrockStepper}(s::Solver{O,S})
+    t  = s.ode.t0
+    dt = s.options.initstep
+    y  = s.ode.y0
+    dy = zero(y)
+
+    J  = Array(eltype(y),length(y),length(y))
+
+    step  = Step(t,copy(y),copy(dy))
+    state = RosenbrockState(step,
+                            dt,
+                            zero(y), # F1
+                            zero(y), # F2
+                            J,       # J
+                            0)       # iters
+    # initialize the derivative and the Jacobian
+    s.ode.F!(t,y,step.dy)
+    s.ode.jac!(t,y,state.J)
+
+    return state
+end
+
+
+function next{O<:ExplicitODE,S<:ModifiedRosenbrockStepper}(s::Solver{O,S}, state)
+
+    stepper = s.stepper
+    ode     = s.ode
+    step    = state.step
+    opts    = s.options
+
+    F1, F2, J = state.F1, state.F2, state.J
+
+    t, dt, y, dy = step.t, state.dt, step.y, step.dy
+    # F!, jac! = ode.F!, ode.jac!
+    d, e32 = stepper.d, stepper.e32
+
+    F0 = dy
+
+    while true
+
+        state.iters += 1
+        if state.iters > s.options.maxiters
+            return ((step.t,step.y), state)
+        end
+
+        # trim the step size to match the bounds of integration
+        dt = min(s.options.tstop-t,dt)
+
+        W = lufact!( eye(J) - dt*d*J )
+
+        # Approximate time-derivative of F, we are using F1 as a
+        # temporary array
+        ode.F!(t+dt/100,y,F1)
+        tder = 100*d*(F1-F0)
+
+        # modified Rosenbrock formula
+        # TODO: allocate some temporary space for these variables
+        k1 = W \ (F0 + tder)
+        ode.F!(t+dt/2, y+dt*k1/2, F1)
+        k2 = W \ (F1 - k1) + k1
+        ynew = y + dt*k2
+        ode.F!(t+dt,   ynew,      F2)
+        k3 = W \ (F2 - e32*(k2 - F1) - 2*(k1 - F0) + tder )
+
+        delta = max(opts.reltol*max(opts.norm(y)::eltype(y),
+                                    opts.norm(ynew)::eltype(y)),
+                    opts.abstol) # allowable error
+
+        err = (dt/6)*(opts.norm(k1 - 2*k2 + k3)::eltype(y))/delta # error estimate
+
+        # upon a failed step decrease the step size
+        dtnew = min(opts.maxstep,
+                    dt*0.8*err^(-1/3) )
+
+        # check if the new solution is acceptable
+        if err <= 1
+
+            # update the state and return
+            step.t     = t+dt
+            state.dt   = dtnew
+            step.y[:]  = ynew
+            step.dy[:] = F2
+            ode.jac!(step.t,step.y,J)
+
+            return ((step.t,step.y), state)
+        else
+            # continue with the decreased time step
+            dt = dtnew
+        end
+
+    end
+
+    return tout, yout
```

Comment:
The term I often see used is `dead code` or `unreachable code`. Sometimes type inference isn't able to determine that controlflow never reaches the end of a function, so you need an extra return statement so that you get better type information in the calling function. (An exception would have the same effect.) Anyway you should include a comment to make the purpouse of the return (or exception) obvious to the reader.